libres/tests/tmpdir.cpp

```diff
@@ -0,0 +1,127 @@
+#include <string>
+#include <optional>
+#include <regex>
+
+#include <unistd.h>
+#include <pwd.h>
+
+#include <catch2/catch.hpp>
+#include <fmt/format.h>
+
+#include "tmpdir.hpp"
+
+using namespace std::string_literals;
+namespace fs = std::filesystem;
+
+namespace {
+/**
+ * Convert string into a nicer directory name
+ *
+ * Downcases, replaces whitespace-ish chars with '_' and removes the
+ * rest.
+ */
+std::string make_dirname(const std::string &dirname) {
+    std::string out;
+    out.reserve(dirname.size());
+
+    for (char c : dirname) {
+        if ('a' <= c && c <= 'z')
+            out.push_back(c);
+        else if ('A' <= c && c <= 'Z')
+            out.push_back(c + 'a' - 'A');
+        else if (c == '_' || c == '-' || c == ' ')
+            out.push_back('_');
+    }
+    return out;
+}
+
+/**
+ * Get POSIX username of the current user
+ *
+ * Falls back to the user's UID if their username is unavailable.
+ */
+std::string get_username() {
+    auto uid = getuid();
+    auto pw = getpwuid(uid);
+    return pw ? std::string{pw->pw_name} : std::to_string(uid);
+}
+
+/**
+ * Create a new directory suffixed with a number
+ *
+ * The prefix is normalised with 'make_dirname'.
+ *
+ * \param basepath Directory in which to create the new directory
+ * \param prefix Prefix of the directory to be created
+ */
+std::filesystem::path make_numbered_dir(const fs::path &basepath,
+                                        const std::string &prefix) {
+    auto norm_prefix = make_dirname(prefix);
+    std::regex re{norm_prefix + "(\\d)+"s};
+    std::smatch match;
+
+    // Find largest entry with the given 'prefix'
+    int number = 0;
+    for (auto entry : fs::directory_iterator(basepath)) {
+        auto entry_str = entry.path().filename().string();
+        if (!std::regex_match(entry_str, match, re))
+            continue;
+        number = std::max(std::stoi(match[1]) + 1, number);
+    }
+
+    // Attempt to create a directory up to 10 times
+    for (int i{}; i < 10; ++i) {
+        std::error_code ec;
+        auto path = basepath / (norm_prefix + std::to_string(number + i));
+        if (fs::create_directory(path, ec)) {
+            // Create a symlink. Error codes are ignored because we don't really care
+            auto current = basepath / (norm_prefix + "current"s);
+            fs::remove(current, ec);
```

Comment:
This will fail due to a race condition if multiple instances of the testsuite runs at once. I think some variant of a lockfile will have to be used.

Reply:
In its current state, we would not be able to safely run test cases which use the TmpDir fixture concurrently.

Reply:
It would also fail if there are multiple jobs running simultaneously on Jenkins

Reply:
Here is pytest implementation of such a lock: https://github.com/pytest-dev/pytest/blob/d6e6b9696d2a2e79cec19e3f7ab497bf34862897/src/_pytest/pathlib.py#L348

Reply:
I think there's a misunderstanding:

1. I'm only deleting the `*current` symlink which points to the newest tmpdir location. This symlink is entirely for the benefit of the programmer and isn't used in the tests. The point of this is that the user can just choose the directory ending in `current` rather than figuring out which number is the largest. If, during a CI run, the removing or symlinking should fail entirely or point to the non-latest test suite, it won't matter. I'm using the error-code, non-throwing versions of the filesystem functions so the code succeeds regardless.
2. `fs::create_directory` is atomic (I think). That is, if I call this function and it succeeds, I know that I am the one who created the directory in question. There could still be a race condition between finding the biggest index and creating the directory. We solve this by having the for-loop counting to 10. Ie, we try to create the next 10 numbers.
3. The locking pytest does is to avoid deleting the temporary data for some old long-running pytest process during cleanup. My code doesn't implement cleanup so there's no need for locking (yet).

Reply:
Sorry, I indeed did not notice that you only cleaned up the symlink.